### PR TITLE
Fill in Genoa tokens

### PR DESCRIPTION
**Pull request by luqmana**
A rough pass at adding the APCB tokens found in Genoa

### CODE REVIEW

File: src/ondisk.rs [hunk 1]
```diff
@@ -8294,20 +8816,70 @@ make_token_accessors! {
...
+        FchI2cI3cSmbusSelect3(default 1, id 0xed13_9de8) | pub get FchI2cI3cSmbusSelect : pub set FchI2cI3cSmbusSelect,
+
+        // TODO: Replaced at some point with APCB_TOKEN_UID_I2C_4 / APCB_TOKEN_UID_I2C_5 ?
+        FchI2cI3cSmbusSelect4(default 1, id 0x2158_efca) | pub get FchI2cI3cSmbusSelect : pub set FchI2cI3cSmbusSelect,
+        FchI2cI3cSmbusSelect5(default 1, id 0x5caa_a37b) | pub get FchI2cI3cSmbusSelect : pub set FchI2cI3cSmbusSelect,
+
+        // TODO: Maybe should be called `FchI2cSdaRxHold` but that's already a WordToken
+        FchIcSdaRxHold(default 0, id 0xa4ba_c3d5) | pub get u8: pub set u8,
+        // Same as `FchI2cSdaHoldOverrideMode` but that's already a WordToken
+        FchI2cSdaHoldOverrideMode2(default 0, id 0x545d_7662) | pub get FchI2cSdaHoldOverrideMode : pub set FchI2cSdaHoldOverrideMode,
```
Comment (daym) [inline src/ondisk.rs:8878]: If we wanted to we could totally remove the direct token interface that takes all the tokens, and for each of them, generates an accessor at the top level with just the token name.

I don't think we have any consumers of that interface anyway.

Not sure what AMD's intention here is. Is it so everyone requires hungarian notation to make sense of the things? ;P

File: src/ondisk.rs [hunk 2]
```diff
@@ -8328,15 +8900,71 @@ make_token_accessors! {
         DfRemapAt1TiB(default 0, id 0x35ee_96f3) | pub get DfRemapAt1TiB : pub set DfRemapAt1TiB,
         DfXgmiTxEqMode(default 0xff, id 0xade7_9549) | pub get DfXgmiTxEqMode : pub set DfXgmiTxEqMode,
         DfInvertDramMap(default 0, id 0x6574_b2c0) | pub get DfToggle : pub set DfToggle,
+        DfXgmiCrcScale(default 5, id 0x5174_f4a0) | pub get u8 : pub set u8,
+        DfCxlMemInterleaving(default 0xFF, id 0x6387_09e6) | pub get DfCxlToggle : pub set DfCxlToggle,
+        DfCxlSublinkInterleaving(default 0xFF, id 0x66f_1625a) | pub get DfCxlToggle : pub set DfCxlToggle,
+        // TODO: Represent bitfield somehow
+        #[cfg_attr(feature = "serde-hex", serde(serialize_with = "SerHex::<StrictPfx>::serialize", deserialize_with = "SerHex::<StrictPfx>::deserialize"))]
+        DfXgmiAcDcCoupledLink(default 0xff, id 0xa7ae_5713) | pub get u8 : pub set u8,
```
Comment (daym) [inline src/ondisk.rs:8908]: Sure, why not? modular_bitfield should work. See MemPmuBistTestSelect for how. For example:

```
        #[derive(BitfieldSpecifier, Debug, PartialEq)]
        #[bits = 1]
        pub enum DfXgmiAcDcMode {
            AcCoupled = 0,
            DcCoupled = 1,
        }

make_bitfield_serde! {
    #[bitfield(bits = 8)]
    #[repr(u8)]
    #[derive(Debug, Copy, Clone, PartialEq)]
    pub struct DfXgmiAcDcCoupledLink {
        pub socket_0_link_0 || #[serde(default)] bool : bool | pub get DfXgmiAcDcMode : pub set DfXgmiAcDcMode,
        pub socket_0_link_1 || #[serde(default)] bool : bool | pub get DfXgmiAcDcMode : pub set DfXgmiAcDcMode,
        pub socket_0_link_2 || #[serde(default)] bool : bool | pub get DfXgmiAcDcMode : pub set DfXgmiAcDcMode,
        pub socket_0_link_3 || #[serde(default)] bool : bool | pub get DfXgmiAcDcMode : pub set DfXgmiAcDcMode,
        pub socket_1_link_0 || #[serde(default)] bool : bool | pub get DfXgmiAcDcMode : pub set DfXgmiAcDcMode,
        pub socket_1_link_1 || #[serde(default)] bool : bool | pub get DfXgmiAcDcMode : pub set DfXgmiAcDcMode,
        pub socket_1_link_2 || #[serde(default)] bool : bool | pub get DfXgmiAcDcMode : pub set DfXgmiAcDcMode,
        pub socket_1_link_3 || #[serde(default)] bool : bool | pub get DfXgmiAcDcMode : pub set DfXgmiAcDcMode,
    }
}
impl_bitfield_primitive_conversion!(DfXgmiAcDcCoupledLink, 0b1111_1111, u8);

```

Or maybe a per-socket struct with bitfieldspecfier, see Ddr4DimmRanks for how.

I can also add it myself if you want.

File: src/ondisk.rs [hunk 3]
```diff
@@ -8468,6 +9107,20 @@ make_token_accessors! {
         // Memory Controller
 
         CbsMemPowerDownDelay(default 0xff, id 0x1ebe_755a) | pub get CbsMemPowerDownDelay : pub set CbsMemPowerDownDelay,
+        #[cfg_attr(feature = "serde-hex", serde(serialize_with = "SerHex::<StrictPfx>::serialize", deserialize_with = "SerHex::<StrictPfx>::deserialize"))]
+        // TODO: would be nice to not have to name this differently
```
Comment (daym) [inline src/ondisk.rs:9111]: We can if we drop one kind of unused API, see other comment here.

**Review by daym — CHANGES_REQUESTED** [review]
Only small nits.

File: src/ondisk.rs [hunk 4]
```diff
@@ -8077,17 +8318,102 @@ impl ToPrimitive for DfCakeCrcThresholdBounds {
...
 #[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
 #[cfg_attr(feature = "serde", serde(deny_unknown_fields))]
 #[cfg_attr(feature = "schemars", derive(schemars::JsonSchema))]
 pub enum MemTrainingHdtControl {
-    DetailedDebugMessages = 5,
-    CoarseDebugMessages = 10,
-    StageCompletionMessages = 200,
-    StageCompletionMessages1 = 201, // FIXME
-    // TODO: Seen in the wild: 201
+    DetailedDebugMessages = 0x04,
```
[the remark this answers is not in the document]
Comment (daym) [reply src/ondisk.rs:8404]: Well, documented is 4 in Genoa and Turin (and nothing in Milan). So 4 it is.

File: src/ondisk.rs (hunk 2 above)
Comment (daym) [reply src/ondisk.rs:8908]: Not sure how many explicit arrays we want here. One?

Comment (luqmana) [reply src/ondisk.rs]: Updated to use bitfield

File: src/ondisk.rs [hunk 5]
```diff
@@ -8294,20 +8816,70 @@ make_token_accessors! {
...
         // Ccx
 
         CcxSevAsidCount(default 1, id 0x5587_6720) | pub get CcxSevAsidCount : pub set CcxSevAsidCount,
+        CcxApicMode(default 0, id 0xf284_ad3f) | pub get CcxApicMode : pub set CcxApicMode,
 
         // Fch
 
         #[cfg_attr(feature = "serde-hex", serde(serialize_with = "SerHex::<StrictPfx>::serialize", deserialize_with = "SerHex::<StrictPfx>::deserialize"))]
-        FchConsoleOutMode(default 0, id 0xddb7_59da) | pub get u8 : pub set u8,
+        FchConsoleOutMode(default 0, id 0xddb7_59da) | pub get FchConsoleOutMode : pub set FchConsoleOutMode,
```
[the remark this answers is not in the document]
Comment (luqmana) [reply src/ondisk.rs:8882]: I'm in favour of revving the version and having a flag day. Especially if we do end up making some API changes. Either way this PR is just into the `issue-110` working branch. Once we're ready to land to main we can update pinned commit in helios and AHIB

**Pull request merged**